sdk/common/src/main/scala/org/typelevel/otel4s/sdk/common/InstrumentationScope.scala

```diff
@@ -0,0 +1,173 @@
+
+  override final def hashCode(): Int =
+    Hash[InstrumentationScope].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: InstrumentationScope =>
+        Hash[InstrumentationScope].eqv(this, other)
+      case _ =>
+        false
+    }
+
+  override final def toString: String =
+    Show[InstrumentationScope].show(this)
+}
+
+object InstrumentationScope {
+
+  /** A builder of [[InstrumentationScope]].
+    */
+  sealed trait Builder {
+
+    /** Sets the version.
+      *
+      * @param version
+      *   the version to set
+      */
+    def setVersion(version: String): Builder
+
+    /** Sets the schema URL.
+      *
+      * @param schemaUrl
+      *   the schema URL to set
+      */
+    def setSchemaUrl(schemaUrl: String): Builder
+
+    /** Sets the attributes.
+      *
+      * @param attributes
+      *   the attributes to set
+      */
+    def setAttributes(attributes: Attributes): Builder
+
+    /** Creates an [[InstrumentationScope]] with the configuration of this
+      * builder.
+      */
+    def build: InstrumentationScope
+  }
+
+  /** An empty [[InstrumentationScope]] */
+  val Empty: InstrumentationScope =
+    apply("", None, None, Attributes.Empty)
+
+  /** Creates a [[Builder]] of [[InstrumentationScope]].
+    *
+    * @param name
+    *   the name of the instrumentation scope
+    */
+  def builder(name: String): Builder =
+    BuilderImpl(name, None, None, Attributes.Empty)
+
+  /** Creates an [[InstrumentationScope]].
+    *
+    * @param name
+    *   the name of the instrumentation scope
+    *
+    * @param version
+    *   the version of the instrumentation scope if the scope has a version
+    *   (e.g. a library version)
+    *
+    * @param schemaUrl
+    *   the Schema URL that should be recorded in the emitted telemetry
+    *
+    * @param attributes
+    *   the instrumentation scope attributes to associate with emitted telemetry
+    */
+  def apply(
+      name: String,
+      version: Option[String],
+      schemaUrl: Option[String],
+      attributes: Attributes
+  ): InstrumentationScope =
+    ScopeImpl(name, version, schemaUrl, attributes)
+
+  implicit val showInstrumentationScope: Show[InstrumentationScope] =
+    Show.show { scope =>
+      show"InstrumentationScope{name=${scope.name}, version=${scope.version}, schemaUrl=${scope.schemaUrl}, attributes=${scope.attributes}}"
+    }
+
+  implicit val hashInstrumentationScope: Hash[InstrumentationScope] =
+    Hash.by { scope =>
+      (scope.name, scope.version, scope.schemaUrl, scope.attributes)
+    }
+
+  private final case class ScopeImpl(
+      name: String,
+      version: Option[String],
+      schemaUrl: Option[String],
+      attributes: Attributes
+  ) extends InstrumentationScope
+
+  private final case class BuilderImpl(
+      name: String,
+      version: Option[String],
+      schemaUrl: Option[String],
+      attributes: Attributes
+  ) extends Builder {
```

Comment:
Actually these can be combined into a single class, then we save an allocation in the `build` method.

Reply:
I don't see reasons why we shouldn't :)